Main menu

thamas opened the pull request:
Task's issue: #49 

Considerations:

- I used classy output as base (see: #70) but 
  - changed file names and class names too to have a usable BEM syntax 
(related info: https://medium.com/integral-vision/drupal-8-twig-add-custom-css-classes-to-menus-based-on-menu-name-35b50142521a)
  - added a button element to manage responsive toggle effect
- used inline svg to display the menu toggle icon and tried to make it accessible 
(related: https://css-tricks.com/accessible-svgs/)
- added a breakpoint which should be tuned when working on the header component.
- I've looked for a simple solution for menu toggle on narrower screens but it has its own limitations. See comment in menu-main.css, line 40 https://github.com/lauriii/umami/compare/master...main-menu#diff-c36a33906a0a7c0777d2171df71adaadR40
- Toggle effect **won't work in the KSS example** because it needs jQuery and Drupal.behaviors. This should be managed in a followup issues: #78 #79
  - see this codepen until the js fix: https://codepen.io/thamas/pen/BwGWGR
- The above JS issue shows _an a11y issue(?)_ of the current toggle solution: if a user has CSS support but has no JS support, she won't be able to access the menu in narrower screens. This should be fixed in a followup issue(?).

thamas committed:
Fix jumping text for now

thamas committed:
#36 Prevent unvanted duplication caused by modifier misuse.

thamas committed:
Set the markup to be as it comes from classy.

thamas committed:
#36 Hide what we do not need and set placeholder text.

thamas committed:
#36 Style placeholder text

thamas committed:
#36 Add search icon and finish styling

thamas committed:
#36 Move visually hidden to base.css

Commit:
View mode highlighted small (#69)

* adding colours to allow us test for ally contrast

* adding tpl file for colours

* first commit of banner block using kss

* first commit of branding block using kss

* logo has been moved to the branding block directory

* first commit of image styles using kss

* first commit of view modes using kss

* initial commit of styling body elements, we should break these into separate files

* move each html text element to its own component

* setting order for typography items

* move body element styling back to base.css, remove body directory from typography directory

* initial commit of form items directory, needed for the field label states

* committing styling for the highlighted small view mode

* small tweaks from pull request feedback

* small tweaks from pull request feedback, volume 2

* finalising hover states

thamas committed:
#49 There is no breadcrumb in the design so delete the related component

thamas committed:
#49 Create basic structure

thamas committed:
#49 Styling and layout for larger screens

thamas committed:
#49 Add some responsive behavior

thamas committed:
#49 Add a media query

tune it later when building the region

thamas committed:
#49 Add toggle button

thamas committed:
Merge branch 'master' into main-menu

* master:
  Search Box and Branding Block (#72)
  View mode highlighted small (#69)

thamas committed:
#49 Add hamburger icon (compressed svg :))

thamas committed:
#49 Menu toggle function for narrower screns

Currently broken in KSS see https://codepen.io/thamas/pen/BwGWGR instead

thamas committed:
#49 Use inline svg instead of background image

lauriii commented on components/components/navigation/menu-main/menu-main.css:
```diff
@@ -0,0 +1,87 @@
+  border-bottom-color: #ff6138;
+}
+
+@media screen and (min-width: 30em) {
```
Would it be possible to add a variation of the responsive navigation so that this could be shown on the desktop?

thamas replied:
@lauriii I'm afraid I do not understand this request… :|

thamas replied:
@lauriii Oh, you mean to see the narrow menu on desktop without changing the width of the browser? Well I could add a narrow container or a class to make the nav narrow but it would be an unneeded thing in the theme. 

I do not know if it is possible to manage it just in the styleguide to have a narrow container just in that without affecting the css added to the theme…

lauriii commented on components/components/navigation/menu-main/menu-main.css:
```diff
@@ -0,0 +1,87 @@
+  text-decoration: none;
+  border-bottom: solid 0.15em transparent;
+  color: inherit;
+  transition: all 0.2s;
```
Would be great to get feedback about this from @kjays since this wasn't something that I was seeing in the original design

thamas replied:
OK. Can we get him to check? :)

thamas replied:
@lauriii Hm, are you talking abou the design of the links? It was in the sketch and in the pdf styleguide too. 

And he provided a "mobile" design on Slack some days ago 
![umami - front - mobile - nav open](https://user-images.githubusercontent.com/1244043/31885760-f410d10a-b7f1-11e7-91fe-3e68f27a9fe4.jpg)
so I updated the code according to it too.

lauriii commented on components/components/navigation/menu-main/menu-main.html.twig:
```diff
@@ -0,0 +1,32 @@
+
+  <ul class="menu-main">
+    <li class="menu-main__item menu-main__item--active-trail">
+      <a href="/" data-drupal-link-system-path="<front>" class="menu-main__link is-active">Home</a>
```
Could we pass the data from a JSON file instead of hardcoding them here?

thamas replied:
Yes, I will do.

lauriii commented on components/components/navigation/menu-main/menu-main.js:
```diff
@@ -0,0 +1,20 @@
+  'use strict';
+  Drupal.behaviors.menutoggle = {
+    attach: function (context, settings) {
+      var toggler = $('.menu-main__toggle');
```
We should use data attributes as selectors in JS

thamas replied:
@lauriii Why? Is there a coding standard about it? Can you point me to that?

thamas replied:
Found two issues:
- https://www.drupal.org/node/1090592
- https://www.drupal.org/node/2346799

lauriii replied:
It is mainly to be able to have documentation that JavaScript expects some element to exist. If you use the classes used for styling the page, the person styling the page assumes they can safely change them to be whatever they would like.

thamas replied:
@lauriii Thanks for letting me know about this! :)

lauriii commented on components/components/navigation/menu-main/menu-main.js:
```diff
@@ -0,0 +1,20 @@
+  Drupal.behaviors.menutoggle = {
+    attach: function (context, settings) {
+      var toggler = $('.menu-main__toggle');
+      var menu = $('.menu-main');
```
We should use data attributes as selectors in JS

lauriii commented on components/components/navigation/menu-main/menu-main.css:
```diff
@@ -0,0 +1,87 @@
+  text-align: left;
+  background-color: transparent;
+}
+.menu-main {
```
I usually try to always place the main element as first in the CSS

thamas replied:
OK

thamas committed:
#49 Change rule order

See: https://github.com/lauriii/umami/pull/80#pullrequestreview-70110011

thamas committed:
#49 Change rule order

See: https://github.com/lauriii/umami/pull/80#discussion_r145328762

thamas committed:
Merge branch 'main-menu' of github.com:lauriii/umami into main-menu

* 'main-menu' of github.com:lauriii/umami:
  #49 Change rule order

thamas committed:
#49 Use json source instead of hardcoded data

See: https://github.com/lauriii/umami/pull/80#discussion_r145328511

thamas committed:
#49 We have to have just .twig as file extension (not .html.twig)!

…to be able to embed the json data into the template

See the previous 9a0a546 commit too

thamas committed:
#49 Update to the (new) narrow screen design

See: https://github.com/lauriii/umami/pull/80#discussion_r145328409

thamas committed:
#49 We should use data attributes as js selectors in Drupal

See:
- https://github.com/lauriii/umami/pull/80#discussion_r145328621 and
- https://github.com/lauriii/umami/pull/80#discussion_r145328634

thamas committed:
#49 Higher specificity is justifiable here

thamas committed:
#49 Typo fix

thamas committed:
#49 Cleanup

thamas commented:
I think it can be merged now! :)
What do you think @lauriii ?

kjays commented:
Looks good though the hamburger icon has no menu-visible styling and presumably needs to switch to a close 'X' when the menu is displayed, separate issue? The Hamburger also has a focus outline that is pretty intrusive.

- Tested in Google Chrome latest OS X
- Tested in Edge 16 on Windows 10
- Tested in IE11 on Windows 10 but styleguide failing to display Examples (separate issue?)

thamas commented:
@kjays Thanks for the review! I think we should let this out and make enhancements in follow-up issues.

- Changing the hamburger icon to a closing X in active state was not part of the design I saw and needs some(?) more coding / different solution(?).
- Focus outline is disturbing I know but hiding it _without affecting accessibility_ is not trivial as I know.
